### ISSUE TITLE
Data specs

### ISSUE THREAD

**Issue by maherr13** [utter-project/fairseq#3]
Hello thanks for your project.

i have a question regarding the data for further finetuning, i need to start collecting data,  what should be the data characteristics ?

could the data have noise ? if not is it ok to run a denoising model ? or it should be 100% noise free
should each wave have one speaker ? should it be like librispeech, each speaker contain few waves in a separate folder? 

if there are any other recommendations, or prevention , please mention them. 

Thanks

**Comment by mzboito** [comment]
Hi,

For fine-tuning, you can use anything you want! 
If you use very clean data, your might be able to have better results with less data. 
If you use noisy data, you might need more, but in the end your model _might_ be more robust to noise.

It really depends on the application you are trying to build.

**Comment by maherr13** [comment]
Thanks for your insights, 

i may need to ask about the clustering phase after training the Hubert [link](https://github.com/utter-project/fairseq/tree/main/examples/hubert/simple_kmeans) 
When I train the Kmeans, should I use the same data the hubert were trained on? I need to use it for Arabic language. Should the data be only Arabic (in the K-means clustering)? Given that it's mHubert, how much variance do I need for good representation (number of speakers), or is it fine also to have the data in a random way?

**Comment by mzboito** [comment]
Hi again,

My answer very quickly became long, sorry about that. I hope it is still understandable. I think the most important thing is to understand what exactly you are trying to do, so that you choose the right approach.

**Continuous pre-training** (sometimes called "fine-tuning", but on self-supervised objective)
Please be aware that if you are **continuous pretraining** the mHuBERT-147 3rd iteration, which means you are restarting training from the last checkpoint with your new data, **you cannot use a new clustering model.** In that setting, the pre-trained model is already over-fitted on the representations from the original clustering, so you should use the [same model](https://huggingface.co/utter-project/mHuBERT-147/blob/main/mhubert147_faiss.index) to produce labels to your new training set. So basically, in this setting the sequence of steps is the following:
1. get your arabic data
2. extract labels using the existing faiss k-means model that I made available (scripts also available at https://github.com/utter-project/mHuBERT-147-scripts/tree/main/03_faiss_indices )
3. restart training using mHuBERT-147 + your data

You only need a new clustering model if you are making a new iteration **from scratch**, or if you decide to reset the vocabulary projection layer on the original model (I never tried this, but it should work). If this is the case, then the rest of my reply is relevant.

**Clustering**
For clustering, your goal should be to create a clustering training set that is the most representative of your fine-tuning training data as possible.
If your fine-tuning Arabic data has several speakers, try to represent that diversity on the clustering training set. 
If you have several datasets, try to represent them following a similar distribution (e.g. if half your data is from CommonVoice, try to sample half of your clustering training data to be from CommonVoice as well).
If your fine-tuning data is small enough and you have machines with large RAM, you might even be able to cluster using the entire fine-tuning training set, which is the ideal solution.

**Comment by maherr13** [comment]
Thanks a lot that is so helpful,

I think I had a wrong understanding of the process, so here is what I was trying to do, **please correct me if I'm wrong**

I wanted to have a model that better represents the Arabic language waves, so I tried to use a multilingual checkpoint (mhubert-147) and have a new k-means trained on my Arabic dataset. so I collected Arabic audio and denoised it using a denoiser model and wanted to train the k-means using it. is that possible so far? what is your recommendation knowing that my target is a model with good discrete representation for the Arabic language?

**Comment by mzboito** [comment]
It is not mandatory to denoise your data, but if you don't have a lot, that might be helpful.

* If you have access to a lot of GPUs:
Train a new iteration.
1. Extract features from your speech data using mHuBERT-147 (https://huggingface.co/utter-project/mHuBERT-147)
2. Train a new k-means model (I recommend faiss-based, as in my paper)
3. Train a new iteration from scratch using your data only

* If you cannot spare the GPUs:
Continuous pretrain the existing 3rd iteration (https://huggingface.co/utter-project/mHuBERT-147)
1. Extract features from your speech data using the 2nd iteration mHuBERT-147 (https://huggingface.co/utter-project/mHuBERT-147-base-2nd-iter)
2. Extract labels using the existing k-means model (https://huggingface.co/utter-project/mHuBERT-147/blob/main/mhubert147_faiss.index)
3. Restart training from the fairseq checkpoint of mHuBERT-147 3rd iteration on the fairseq fork (https://huggingface.co/utter-project/mHuBERT-147/blob/main/checkpoint_best.pt)
